src/activity_handlers.cpp

```diff
@@ -1339,19 +1339,21 @@ void activity_handlers::train_finish( player_activity *act, player *p )
     const skill_id sk( act->name );
     if( sk.is_valid() ) {
         const Skill &skill = sk.obj();
+        std::string skill_name = skill.name();
         int new_skill_level = p->get_skill_level( sk ) + 1;
         p->set_skill_level( sk, new_skill_level );
-        add_msg(m_good, _("You finish training %s to level %d."),
-                skill.name().c_str(),
-                new_skill_level);
+        add_msg( m_good, _( "You finish training %s to level %d." ), skill_name, new_skill_level );
         if( new_skill_level % 4 == 0 ) {
             //~ %d is skill level %s is skill name
             p->add_memorial_log(pgettext("memorial_male", "Reached skill level %1$d in %2$s."),
                                 pgettext("memorial_female", "Reached skill level %1$d in %2$s."),
-                                new_skill_level, skill.name().c_str());
+                                new_skill_level, skill_name );
         }
-
-        lua_callback("on_skill_increased");
+        CallbackArgumentContainer lua_callback_args_info;
+        lua_callback_args_info.emplace_back( CallbackArgument( "skill_increased_source", "training" ) );
```

Comment:
When you use one of the `emplace` functions of a standard container, you don't need to create an object of the contained type. The `emplace` functions construct such an object on their own and forward *all* there arguments to its constructor.

In short: `some_container.emplace_back( 1, 2, 3, 4 )` is (mostly) the same as `some_container.push_back( Type( 1, 2, 3, 4 ) )`.